docs/development/architecture.org

```diff
@@ -0,0 +1,130 @@
+Nix-Gui may be considered to have two components: the backend and the frontend.
+
+The backend (=nixui/options=) reads information from both NixOS and the configuration path being edited, then converts them to data-structures useful for navigation, search, and editing. The backend maintains in-memory state for all changes and commits those changes to disk by writing to the configuration path.
+
+The frontend (=nixui/graphics=) constructs various widgets comprising Nix-Gui and loads them with data from the backend. The frontend doesn't manage any edit state. All "edit" actions performed in the GUI send a request-for-update to the backend.
+
+* Backend
+The backend handles three major tasks: configuration retrieval to initialize data-structures, in-memory configuration mutation, and configuration commitment. It provides APIs for frontend interaction, but doesn't concern itself with how the frontend renders any of its data.
+
+** Loading the Option Tree Data-Structure
+Nix-Gui queries information from the configuration path and =<nixpkgs>= in order to construct a data-structure (=nixui.options.option_tree.OptionTree=) mapping options and attributes to metadata (=nixui.options.option_tree.OptionData=) including the attributes =NixType= and =Optiondefinition=:
+- =_type=: A static python-object representation of the option type (=NixType=). Informs the frontend of how =FieldWidget= should be generated and input checking should be performed.
+- =system_default_definition=: Provided by =<nixpkgs>=, the default value if nothing is set in the configuration path.
+- =configured_definition=: Provided by the configuration path, overrides =system_default_definition=.
+- =in_memory_definition=: Undefined when the data-structure is loaded, can be mutated by the user and later committed to disk.
+
+=nixui/nix/lib.nix= (invoked by =nixui.options.nix_eval=) contains the function =get_all_nixos_options()= which is used to retrieve =OptionData= including =system_default_definition= and =_type= for the universe of NixOS options.
+
+=nixui/options/parser.py= contains the function =get_all_option_values(module_path)= which
+- retrieves all =attributePath = string representation of nix expression;= pairs in the file by parsing the =nixui.options.syntax_tree.SyntaxTree= constructed using =nix_dump_syntax_tree_json=.
+- resolves the path of all =imports= and recurses
+
+The result of =get_all_option_values(module_path)= is used to set =OptionData.configured_definition= where applicable.
+
+*** Caveat
+The =OptionTree= has a node for each option in the option tree, as well as attributes that aren't actually options.
+
+For example =users.ldap.loginPam= is a valid nix option. However its "parent", =users.ldap= isn't a valid option. Regardless, =users.ldap= has its own node in the =OptionTree=.
```

Comment:
> However its "parent", =users.ldap= isn't a valid option. Regardless, =users.ldap= has its own node in the =OptionTree=.

wdym "isn't a valid option"?

Reply:
`users.ldap` isn't defined by `mkOption`, it's an attribute set containing options defined by `mkOption`. Is there a better way to word this you could suggest?

Reply:
branch nodes vs leaf nodes?

constant branch nodes vs variable branch nodes?

Reply:
Problem is that options such as `attribute set of ...` and `list of ...` are also branch nodes. The `OptionTree` contains options, ancestors of options, and attributes / elements of options.

Reply:
to make it simple, we have paths and values: `users.ldap.home = "/var/run/ldap"`
every path is a list of keys: users, ldap, home

some keys are constant: users, home (as allowed by the config schema)
other keys are variable: ldap

currently, how does the frontend know if an existing key is constant or varaible?
and, how are variable keys declared in the schema?

Reply:
Static attributes and mutable attributes maybe?

>and, how are variable keys declared in the schema?

`users.ldap.base` is a key in the `dict` returned by `get_all_nixos_options()`, 

`OptionTree` creates nodes for `users.ldap.base`, `users.ldap`, and `users`. It labels each of these nodes with `OptionData`.

`users.ldap.home` has `OptionData._type=StringType()`. However static attributes such as `users.ldap` default to `AttrsType()`.

>currently, how does the frontend know if an existing key is constant or varaible?

The frontend considers the type of the attribute to determine whether it's static or mutable (or in `navlist.py` "dynamic") https://github.com/nix-gui/nix-gui/blob/master/nixui/graphics/navlist.py#L15-L20
The "else" case here is for `AttrsType`.

Reply:
> It labels each of these nodes with OptionData.

wdym "label"? i guess you mean:
node data is stored in `node.data` and the type is in `node.data._type`

> https://github.com/nix-gui/nix-gui/blob/master/nixui/graphics/navlist.py#L15-L20

`types.AttrsOfType` and `types.ListOfType` should be renamed ...

why is there `StaticAttrsOf` but no `StaticListOf`?

for example [networking.firewall.allowedTCPPorts](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/firewall.nix#L259) should be a `StaticListOf` no?

> in navlist.py "dynamic"

dynamic sounds like something that is auto-generated ...
can we say `constant key` vs `variable key`?

variable key: must be set by the user, usually a name for something
constant key: must be valid according to the schema

to clarify, by "key" i mean one item in the attrpath / keypath

```nix
{
  someConfig.someModule."this.is.one.key"."another/key".someProp = "some value";
# ^          ^          ^                 ^             ^          ^
# attrset    key1       key2              key3          key4       value
}
```

Reply:
>types.AttrsOfType and types.ListOfType should be renamed ...

Indeed

>why is there StaticAttrsOf but no StaticListOf?

>for example networking.firewall.allowedTCPPorts should be a StaticListOf no?

Is there a way we can determine that there are a fixed number of ports for this option using nix? If so, then yes. Otherwise, no.

>dynamic sounds like something that is auto-generated ...
>can we say constant key vs variable key?

I think variable is overloaded, so I suggest constant vs mutable.

Also the key isn't what's mutated for `someConfig.someModule`, its definition is.

`someConfig.someModule."this.is.one.key"` is a "mutable key", but when we're determining how to handle an option, the fact that it's a mutable key is irrelevant. What matters is if a member of the set has a mutable key. Therefore the question is whether the attribute itself (not its path) is mutable.

IMO "constant attribute" and "mutable attribute" make more sense.

Reply:
* schema-defined attribute vs user-defined attribute (nix says attribute, long and correct)
* schema attribute vs user attribute (nix says attribute)
* schema key vs user key (short and sweet)
* schema vs user

would be more clear in this context

>> why is there StaticAttrsOf but no StaticListOf?
>
>> for example networking.firewall.allowedTCPPorts should be a StaticListOf no?
>
>Is there a way we can determine that there are a fixed number of ports for this option using nix? If so, then yes. Otherwise, no.

you talk about user-defined values
i talk about user-defined keys

but probably there are no "user-named" lists in the nixos config schema

```nix
{
  someConfig.someModule."this-is-my-list" = [ 1 2 3 4 ];
}
```

and only "schema-named" lists

```nix
{
  someConfig.someModule."this-is-my-set".someList = [ 1 2 3 4 ];
}
```

Reply:
>schema-defined attribute vs user-defined attribute (nix says attribute, long and correct)
>schema attribute vs user attribute (nix says attribute)
>schema key vs user key (short and sweet)
>schema vs user

Mhm, I could write a clarifying bit that explains the schema attributes are defined by the `<nixos>` option declarations, and the user attributes are defined through user definitions.

>but probably there are no "user-named" lists in the nixos config schema

Only a handful, e.g. `services.prometheus.scrapeConfigs."*".params`. It is `attribute set of list of stringss` type.

Reply:
@milahu Does this clarify? https://github.com/nix-gui/nix-gui/pull/214/commits/5c3d741dcbe8a1c811c449817710cf8c6cd0a71f